Automatic Joiners removal in Commissioner

rlubos opened the pull request:
This PR resolves #1377.

Commissioner will now remove Joiners after successfull commissioning or after a timeout period.

Again, I've hit the RAM shortage on CC2538, hence I did some memory savings in JoinerRouter class. 8 bytes only, but it was enough.

nibanks commented on include/openthread/commissioner.h:
```diff
@@ -243,6 +243,25 @@ OTAPI ThreadError OTCALL otCommissionerGeneratePSKc(otInstance *aInstance, const
+ * @param[in]  aTimeout   A time after which Joiners are automatically removed, in seconds.
+ *
+ */
+OTAPI void OTCALL otCommissionerSetJoinerTimeout(otInstance *aInstance, uint32_t aTimeout);
```
What if we just made this timeout parameters a argument of the `otCommissionerAddJoiner` API? Then you wouldn't have to save the `uint32_t mJoinerTimeout` in the commissioner, saving those bytes.

jwhui replied:
I agree.  It would also make it more natural to specify a different timeout for each joiner (if there's a use case for that).

rlubos replied:
Sound good :+1: , I'll refactor it after weekend.

rlubos commented:
I've updated the otCommissionerAddJoiner API to support individial timeout setting, and updated the CLI (though I've made the parameter optional). Also updated timer management on the Commissioner side.
@nibanks I see that Appveyor failed after API update. Could you provide the information what and where should be updated in Windows files?

nibanks commented:
@rlubos, you need to update the Windows user-mode implementation of the API [here](https://github.com/NordicSemiconductor/openthread/blob/465dac55ead18c46d4fc145dda813cf44cfc237d/examples/drivers/windows/otApi/otApi.cpp#L3501). That will entail the following:

1. Update the signature of the function for the new input.
2. Update the function itself to encode the variable in the buffer being sent to kernel mode.
3. Update the comment for `IOCTL_OTLWF_OT_COMMISIONER_ADD_JOINER` to indicate the new variable.
4. Decode the new variable when processing the `IOCTL` [here](https://github.com/NordicSemiconductor/openthread/blob/465dac55ead18c46d4fc145dda813cf44cfc237d/examples/drivers/windows/otLwf/iocontrol.c#L5581).
5. After decoding, pass the new variable to the actual `otCommissionerAddJoiner` call.

nibanks commented on examples/drivers/windows/otApi/otApi.cpp:
```diff
@@ -3514,13 +3515,14 @@ otCommissionerAddJoiner(
     if (aExtAddressValid)
         memcpy_s(Buffer + sizeof(GUID) + sizeof(uint8_t), sizeof(Buffer) - sizeof(GUID) - sizeof(uint8_t), aExtAddress, sizeof(otExtAddress));
     memcpy_s(Buffer + sizeof(GUID) + sizeof(uint8_t) + sizeof(otExtAddress), sizeof(Buffer) - sizeof(GUID) - sizeof(uint8_t) - sizeof(otExtAddress), aPSKd, aPSKdLength);
+    memcpy_s(Buffer + sizeof(GUID) + sizeof(uint8_t) + sizeof(otExtAddress) + aPSKdLength + 1, sizeof(Buffer) - sizeof(GUID) - sizeof(uint8_t) - sizeof(otExtAddress) - aPSKdLength - 1, aTimeout, sizeof(aTimeout));
```
You need an `&` before `aTimeout`.

nibanks commented on examples/drivers/windows/otLwf/iocontrol.c:
```diff
@@ -5595,19 +5595,20 @@ otLwfIoCtl_otCommissionerAddJoiner(
             uint8_t aExtAddressValid = *(uint8_t*)InBuffer;
             const otExtAddress *aExtAddress = aExtAddressValid == 0 ? NULL : (otExtAddress*)(InBuffer + sizeof(uint8_t));
             char *aPSKd = (char*)(InBuffer + sizeof(uint8_t) + sizeof(otExtAddress));
+            uint32_t aTimeout = *(uint32_t*)(InBuffer + sizeof(uint8_t) + sizeof(otExtAddress) + aPSKdBufferLength;
```
Missing a `)` at the end of the line.

rlubos commented:
@nibanks Thanks for detailed instructions, hope it works now!

nibanks commented:
Looks like you need to update [otNodeApi.cpp](https://github.com/NordicSemiconductor/openthread/blob/4bacae786b96e58f9baf859d9aa8cbd377a754c5/examples/drivers/windows/otNodeApi/otNodeApi.cpp#L1008)'s usage of the API. Just pass the a default value in probably.

rlubos commented:
Ok, I've followed your advice, finally AppVeyor is green.
